### ISSUE TITLE
FromSql error handling not compatible with TypedRelationalValueBufferFactoryFactory

### ISSUE THREAD

**Issue by roji** [aspnet/EntityFramework#6563]
The test `Bad_data_error_handling_invalid_cast` fails on Npgsql, which uses `TypedRelationalValueBufferFactoryFactory` instead of the default `UntypedRelationalValueBufferFactoryFactory`.

First, reading `SupplierID` (an integer) as a `ProductName` (a string) works, since Npgsql allows reading casting int fields to strings (i.e. when `NpgsqlDataReader.GetString()` is called). `ProductName`, which is read as `SupplierID`, would throw an exception except that in this test it's a navigation and so is never accessed. It's sufficient to switch around `ProductName` with `UnitsInStock` instead and an exception is generated.

However, the error isn't caught in `EntityMaterializerSource.ThrowReadValueException` and rethrown as the proper expected exception, since when `TypedRelationalValueBufferFactoryFactory` is used the exception is thrown earlier, when the ValueBuffer is constructed (QueryMethodProvider.cs:42) rather than when it is accessed from the shaper (QueryMethodProvider.cs:48).

**Comment by ErikEJ** [comment]
Related to https://github.com/aspnet/EntityFramework/issues/6471 ?

**Comment by rowanmiller** [comment]
We should update the test as @roji suggests

**Comment by roji** [comment]
This is still not fixed. The Npgsql provider, which uses `TypedRelationalValueBufferFactoryFactory`, fails on `Bad_data_error_handling_invalid_cast_no_tracking` and `Bad_data_error_handling_invalid_cast`. Once I change Npgsql to use `UntypedRelationalValueBufferFactoryFactory` the tests pass.

**Comment by anpete** [comment]
@roji How does the test fail? This is strange because we should now be adding the exception handling around all calls to the reader. Are you able to post a query plan for one of the failing tests?

**Comment by roji** [comment]
@anpete well, it seems that an exception is expected but simply isn't thrown - I haven't dived deeper into it. You should be able to easily reproduce this by cloning [Npgsql.EntityFrameworkCore.PostgreSQL](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL), removing the skips in FromSqlQueryNpgsqlTest, and running a test as below (let me know if I can provide any other details).

```
C:\projects\Npgsql.EntityFrameworkCore.PostgreSQL\test\Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests>dotnet test -f net452 --filter "FullyQualifiedName=Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests.FromSqlQueryNpgsqlTest.Bad_data_error_handling_invalid_cast"
Build started, please wait...
Build completed.

Test run for C:\projects\Npgsql.EntityFrameworkCore.PostgreSQL\test\Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests\bin\Debug\net452\Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests.dll(.NETFramework,Version=v4.5.2)
Microsoft (R) Test Execution Command Line Tool Version 15.0.0.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
[xUnit.net 00:00:01.2976331]   Discovering: Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests
[xUnit.net 00:00:05.0840805]   Discovered:  Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests
[xUnit.net 00:00:06.2308507]   Starting:    Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests
[xUnit.net 00:00:08.5745145]     Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests.FromSqlQueryNpgsqlTest.Bad_data_error_handling_invalid_cast [FAIL]
[xUnit.net 00:00:08.5752750]       Assert.Throws() Failure
[xUnit.net 00:00:08.5753584]       Expected: typeof(System.InvalidOperationException)
[xUnit.net 00:00:08.5754018]       Actual:   (No exception was thrown)
[xUnit.net 00:00:08.5761240]       Stack Trace:
[xUnit.net 00:00:08.5772732]            at Microsoft.EntityFrameworkCore.Specification.Tests.FromSqlQueryTestBase`1.Bad_data_error_handling_invalid_cast()
[xUnit.net 00:00:08.5913343]   Finished:    Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests
Failed   Npgsql.EntityFrameworkCore.PostgreSQL.FunctionalTests.FromSqlQueryNpgsqlTest.Bad_data_error_handling_invalid_cast
Error Message:
 Assert.Throws() Failure
Expected: typeof(System.InvalidOperationException)
Actual:   (No exception was thrown)
Stack Trace:
   at Microsoft.EntityFrameworkCore.Specification.Tests.FromSqlQueryTestBase`1.Bad_data_error_handling_invalid_cast()

Total tests: 1. Passed: 0. Failed: 1. Skipped: 0.
Test Run Failed.
Test execution time: 8.9436 Seconds
```

**Comment by smitpatel** [comment]
@anpete @roji - I was able to get above test failure in local machine. I am having hard time debugging with pre-release bits of CLI/SDK & Asp.Net nightly packages. @anpete - would you like to take a look?

**Comment by smitpatel** [comment]
@roji - I debugged into this and found out root cause. There are 3 test failures. 2 of them fails with no exception where it is supposed to give exception. I tried with ADO.NET provider for postgre, when we call `NpgSqlDataReader.GetString` regardless of the type of the column, if it can be converted to string (like int type) the conversion will happen. SqlServer is much stricter about this kind of type matching. The tests in question tries to fetch string value from integer column and throws exception for SqlServer. SQLite also allows type conversions internally so we have disabled those tests there. I will update test so that it tries to get integer value from string column which throws exception for PostgreSQL too.

There is 3rd test failure which throws different exception message which can be probably the random mix of packages I have accumulated in my machine.
